File: assets/js/apple-music-jwt.py
#!/usr/bin/env python3
"""
Apple Music JWT Token Generator
Generates properly formatted JWT tokens for Apple Music API authentication
"""

import jwt
import time
import json
from datetime import datetime, timedelta
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec
import requests


class AppleMusicJWT:
    """Apple Music JWT token generator with proper ES256 signing"""
    
    def __init__(self, team_id, key_id, private_key_path=None, private_key_content=None):
        """
        Initialize Apple Music JWT generator
        
        Args:
            team_id (str): 10-character Team ID from Apple Developer account
            key_id (str): 10-character Key ID from MusicKit private key
            private_key_path (str): Path to .p8 private key file
            private_key_content (str): Private key content as string
        """
        self.team_id = team_id
        self.key_id = key_id
        
        if private_key_path:
            with open(private_key_path, 'r') as f:
                self.private_key = f.read()
        elif private_key_content:
            self.private_key = private_key_content
        else:
            raise ValueError("Either private_key_path or private_key_content must be provided")
        
        self._validate_inputs()
    
    def _validate_inputs(self):
        """Validate required inputs"""
        if not self.team_id or len(self.team_id) != 10:
            raise ValueError("Team ID must be exactly 10 characters")
        
        if not self.key_id or len(self.key_id) != 10:
            raise ValueError("Key ID must be exactly 10 characters")
        
        if not self.private_key:
            raise ValueError("Private key is required")
    
    def generate_token(self, expiration_hours=24, origins=None):
        """
        Generate a properly formatted JWT token for Apple Music API
        
        Args:
            expiration_hours (int): Token expiration in hours (default: 24, max: 4320 = 6 months)
            origins (list): Optional list of allowed origins for web apps
            
        Returns:
            str: JWT token
        """
        # Calculate timestamps
        now = int(time.time())
        max_expiration = now + 15777000  # 6 months max per Apple
        requested_expiration = now + (expiration_hours * 3600)
        expiration = min(requested_expiration, max_expiration)
        
        # JWT Header - MUST use ES256 algorithm
        headers = {
            "alg": "ES256",     # Required: ECDSA using P-256 and SHA-256
            "kid": self.key_id  # Required: 10-character Key ID
        }
        
        # JWT Payload (Claims)
        payload = {
            "iss": self.team_id,  # Required: 10-character Team ID
            "iat": now,           # Required: Issued at time (Unix timestamp)
            "exp": expiration     # Required: Expiration time (Unix timestamp)
        }
        
        # Add origin claim for web applications (optional but recommended)
        if origins and isinstance(origins, list) and len(origins) > 0:
            payload["origin"] = origins
        
        try:
            # Generate the token using PyJWT with ES256 algorithm
            token = jwt.encode(
                payload=payload,
                key=self.private_key,
                algorithm="ES256",
                headers=headers
            )
            
            return token
        except Exception as e:
            raise Exception(f"Failed to generate JWT token: {str(e)}")
    
    def validate_token_format(self, token):
        """
        Validate token format and claims without verifying signature
        
        Args:
            token (str): JWT token to validate
            
        Returns:
            dict: Decoded token information
        """
        try:
            # Decode without verification to check format
            header = jwt.get_unverified_header(token)
            payload = jwt.decode(token, options={"verify_signature": False})
            
            # Validate required fields
            if header.get('alg') != 'ES256':
                raise ValueError("Invalid algorithm: must be ES256")
            
            if not header.get('kid') or len(header.get('kid')) != 10:
                raise ValueError("Invalid Key ID: must be 10 characters")
            
            if not payload.get('iss') or len(payload.get('iss')) != 10:
                raise ValueError("Invalid Team ID: must be 10 characters")
            
            if not payload.get('iat') or not payload.get('exp'):
                raise ValueError("Missing required claims: iat and exp are required")
            
            now = int(time.time())
            if payload.get('exp') <= now:
                raise ValueError("Token has expired")
            
            if payload.get('iat') > now + 60:  # Allow 60 seconds clock skew
                raise ValueError("Token issued in the future")
            
            return {
                'header': header,
                'payload': payload,
                'is_valid': True,
                'expires_at': datetime.fromtimestamp(payload.get('exp')),
                'issued_at': datetime.fromtimestamp(payload.get('iat'))
            }
        except Exception as e:
            raise ValueError(f"Token validation failed: {str(e)}")


class AppleMusicAuth:
    """Helper class for Apple Music API authentication"""
    
    @staticmethod
    def create_token(team_id, key_id, private_key_path=None, private_key_content=None, **options):
        """
        Create a new JWT token for Apple Music API
        
        Args:
            team_id (str): 10-character Team ID
            key_id (str): 10-character Key ID
            private_key_path (str): Path to private key file
            private_key_content (str): Private key content
            **options: Additional options (expiration_hours, origins)
            
        Returns:
            str: JWT token
        """
        jwt_generator = AppleMusicJWT(
            team_id=team_id,
            key_id=key_id,
            private_key_path=private_key_path,
            private_key_content=private_key_content
        )
        
        return jwt_generator.generate_token(
            expiration_hours=options.get('expiration_hours', 24),
            origins=options.get('origins')
        )
    
    @staticmethod
    def test_connection(token):
        """
        Test Apple Music API connection with generated token
        
        Args:
            token (str): JWT token
            
        Returns:
            dict: API response
        """
        try:
            headers = {
                'Authorization': f'Bearer {token}',
                'Content-Type': 'application/json'
            }
            
            # Test with a simple catalog request
            response = requests.get(
                'https://api.music.apple.com/v1/catalog/us/songs/203709340',
                headers=headers,
                timeout=10
            )
            
            if not response.ok:
                raise Exception(f"API request failed: {response.status_code} - {response.text}")
            
            return response.json()
        except Exception as e:
            raise Exception(f"Connection test failed: {str(e)}")


def main():
    """Example usage and testing"""
    import os
    
    # Example configuration (replace with your actual values)
    TEAM_ID = os.getenv('APPLE_TEAM_ID', 'YOUR_TEAM_ID')  # 10 characters
    KEY_ID = os.getenv('APPLE_KEY_ID', 'YOUR_KEY_ID')     # 10 characters
    PRIVATE_KEY_PATH = os.getenv('APPLE_PRIVATE_KEY_PATH', 'path/to/your/AuthKey_XXXXXXXXXX.p8')
    
    try:
        # Generate token
        print("Generating Apple Music JWT token...")
        token = AppleMusicAuth.create_token(
            team_id=TEAM_ID,
            key_id=KEY_ID,
            private_key_path=PRIVATE_KEY_PATH,
            expiration_hours=24,
            origins=['https://yourdomain.com']  # Optional for web apps
        )
        
        print(f"✅ Token generated successfully!")
        print(f"Token length: {len(token)} characters")
        print(f"Token: {token[:50]}...")
        
        # Validate token format
        jwt_generator = AppleMusicJWT(TEAM_ID, KEY_ID, private_key_path=PRIVATE_KEY_PATH)
        token_info = jwt_generator.validate_token_format(token)
        print(f"✅ Token validation passed!")
        print(f"Expires at: {token_info['expires_at']}")
        print(f"Issued at: {token_info['issued_at']}")
        
        # Test API connection
        print("\nTesting Apple Music API connection...")
        result = AppleMusicAuth.test_connection(token)
        print(f"✅ API connection successful!")
        print(f"Test song: {result.get('data', [{}])[0].get('attributes', {}).get('name', 'Unknown')}")
        
    except Exception as e:
        print(f"❌ Error: {str(e)}")
        return 1
    
    return 0


if __name__ == "__main__":
    exit(main())

File: assets/js/apple-music-jwt.js
/**
 * Apple Music JWT Token Generator
 * Generates properly formatted JWT tokens for Apple Music API authentication
 */

class AppleMusicJWT {
    constructor(teamId, keyId, privateKey) {
        this.teamId = teamId;     // 10-character Team ID from Apple Developer account
        this.keyId = keyId;       // 10-character Key ID from your MusicKit private key
        this.privateKey = privateKey; // Private key content (.p8 file)
    }

    /**
     * Generate a properly formatted JWT token for Apple Music API
     * @param {number} expirationHours - Token expiration in hours (default: 24, max: 4320 = 6 months)
     * @param {string[]} origins - Optional array of allowed origins for web apps
     * @returns {string} JWT token
     */
    async generateToken(expirationHours = 24, origins = null) {
        // Validate inputs
        if (!this.teamId || this.teamId.length !== 10) {
            throw new Error('Team ID must be exactly 10 characters');
        }
        if (!this.keyId || this.keyId.length !== 10) {
            throw new Error('Key ID must be exactly 10 characters');
        }
        if (!this.privateKey) {
            throw new Error('Private key is required');
        }

        // Calculate timestamps
        const now = Math.floor(Date.now() / 1000);
        const maxExpiration = now + (15777000); // 6 months max per Apple
        const requestedExpiration = now + (expirationHours * 3600);
        const expiration = Math.min(requestedExpiration, maxExpiration);

        // JWT Header - MUST use ES256 algorithm
        const header = {
            alg: "ES256",  // Required: ECDSA using P-256 and SHA-256
            kid: this.keyId // Required: 10-character Key ID
        };

        // JWT Payload (Claims)
        const payload = {
            iss: this.teamId,    // Required: 10-character Team ID
            iat: now,            // Required: Issued at time (Unix timestamp)
            exp: expiration      // Required: Expiration time (Unix timestamp)
        };

        // Add origin claim for web applications (optional but recommended)
        if (origins && Array.isArray(origins) && origins.length > 0) {
            payload.origin = origins;
        }

        try {
            // Import the private key for ES256 signing
            const privateKeyObject = await this.importPrivateKey(this.privateKey);
            
            // Create and sign the JWT
            const token = await this.signJWT(header, payload, privateKeyObject);
            
            return token;
        } catch (error) {
            throw new Error(`Failed to generate JWT token: ${error.message}`);
        }
    }

    /**
     * Import private key for ES256 signing
     * @param {string} privateKeyPem - Private key in PEM format
     * @returns {CryptoKey} Imported private key
     */
    async importPrivateKey(privateKeyPem) {
        // Clean the private key format
        let cleanKey = privateKeyPem
            .replace(/-----BEGIN PRIVATE KEY-----/, '')
            .replace(/-----END PRIVATE KEY-----/, '')
            .replace(/\s/g, '');

        try {
            // Convert base64 to ArrayBuffer
            const keyData = Uint8Array.from(atob(cleanKey), c => c.charCodeAt(0));
            
            // Import the key for ES256 signing
            return await crypto.subtle.importKey(
                'pkcs8',
                keyData,
                {
                    name: 'ECDSA',
                    namedCurve: 'P-256'
                },
                false,
                ['sign']
            );
        } catch (error) {
            throw new Error(`Invalid private key format: ${error.message}`);
        }
    }

    /**
     * Sign JWT using ES256 algorithm
     * @param {object} header - JWT header
     * @param {object} payload - JWT payload
     * @param {CryptoKey} privateKey - Private key for signing
     * @returns {string} Signed JWT token
     */
    async signJWT(header, payload, privateKey) {
        // Encode header and payload
        const encodedHeader = this.base64UrlEncode(JSON.stringify(header));
        const encodedPayload = this.base64UrlEncode(JSON.stringify(payload));
        
        // Create signing input
        const signingInput = `${encodedHeader}.${encodedPayload}`;
        const signingInputBuffer = new TextEncoder().encode(signingInput);
        
        // Sign with ES256
        const signature = await crypto.subtle.sign(
            {
                name: 'ECDSA',
                hash: 'SHA-256'
            },
            privateKey,
            signingInputBuffer
        );
        
        // Encode signature
        const encodedSignature = this.base64UrlEncode(signature);
        
        // Return complete JWT
        return `${signingInput}.${encodedSignature}`;
    }

    /**
     * Base64 URL-safe encoding
     * @param {string|ArrayBuffer} data - Data to encode
     * @returns {string} Base64 URL-safe encoded string
     */
    base64UrlEncode(data) {
        let base64;
        
        if (typeof data === 'string') {
            base64 = btoa(data);
        } else {
            // ArrayBuffer
            const bytes = new Uint8Array(data);
            let binary = '';
            for (let i = 0; i < bytes.byteLength; i++) {
                binary += String.fromCharCode(bytes[i]);
            }
            base64 = btoa(binary);
        }
        
        // Convert to URL-safe format
        return base64
            .replace(/\+/g, '-')
            .replace(/\//g, '_')
            .replace(/=/g, '');
    }

    /**
     * Validate token format and claims
     * @param {string} token - JWT token to validate
     * @returns {object} Decoded token information
     */
    validateTokenFormat(token) {
        const parts = token.split('.');
        if (parts.length !== 3) {
            throw new Error('Invalid JWT format: must have 3 parts');
        }

        try {
            const header = JSON.parse(atob(parts[0].replace(/-/g, '+').replace(/_/g, '/')));
            const payload = JSON.parse(atob(parts[1].replace(/-/g, '+').replace(/_/g, '/')));
            
            // Validate required fields
            if (header.alg !== 'ES256') {
                throw new Error('Invalid algorithm: must be ES256');
            }
            if (!header.kid || header.kid.length !== 10) {
                throw new Error('Invalid Key ID: must be 10 characters');
            }
            if (!payload.iss || payload.iss.length !== 10) {
                throw new Error('Invalid Team ID: must be 10 characters');
            }
            if (!payload.iat || !payload.exp) {
                throw new Error('Missing required claims: iat and exp are required');
            }
            
            const now = Math.floor(Date.now() / 1000);
            if (payload.exp <= now) {
                throw new Error('Token has expired');
            }
            if (payload.iat > now + 60) { // Allow 60 seconds clock skew
                throw new Error('Token issued in the future');
            }
            
            return { header, payload, isValid: true };
        } catch (error) {
            throw new Error(`Token validation failed: ${error.message}`);
        }
    }
}

// Usage example and helper functions
const AppleMusicAuth = {
    /**
     * Create a new JWT token for Apple Music API
     * @param {string} teamId - 10-character Team ID
     * @param {string} keyId - 10-character Key ID  
     * @param {string} privateKey - Private key content
     * @param {object} options - Additional options
     * @returns {Promise<string>} JWT token
     */
    async createToken(teamId, keyId, privateKey, options = {}) {
        const {
            expirationHours = 24,
            origins = null
        } = options;
        
        const jwt = new AppleMusicJWT(teamId, keyId, privateKey);
        return await jwt.generateToken(expirationHours, origins);
    },

    /**
     * Test Apple Music API connection with generated token
     * @param {string} token - JWT token
     * @returns {Promise<object>} API response
     */
    async testConnection(token) {
        try {
            const response = await fetch('https://api.music.apple.com/v1/catalog/us/songs/203709340', {
                headers: {
                    'Authorization': `Bearer ${token}`,
                    'Content-Type': 'application/json'
                }
            });
            
            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`API request failed: ${response.status} - ${errorText}`);
            }
            
            return await response.json();
        } catch (error) {
            throw new Error(`Connection test failed: ${error.message}`);
        }
    }
};

/**
 * Apple Music Service with Netlify Integration
 * Uses environment configuration to generate tokens automatically
 */
class AppleMusicService {
    constructor() {
        this.token = null;
        this.tokenExpiry = null;
        this.credentials = {
            teamId: '5S855HB895',  // Your Apple Team ID
            keyId: 'F9Q8YRKX3T',   // Your Apple Key ID
            privateKey: null       // Will be loaded from config
        };
    }

    /**
     * Initialize service with credentials from environment config
     */
    async initialize() {
        try {
            // Wait for environment config to be ready
            if (!window.envConfig?.initialized) {
                await new Promise((resolve, reject) => {
                    const timeout = setTimeout(() => reject(new Error('Config timeout')), 5000);
                    
                    window.addEventListener('configReady', () => {
                        clearTimeout(timeout);
                        resolve();
                    }, { once: true });
                    
                    window.addEventListener('configError', (e) => {
                        clearTimeout(timeout);
                        reject(new Error(e.detail.error));
                    }, { once: true });
                });
            }

            // Try to get Apple Music private key from Netlify function
            await this.loadPrivateKeyFromNetlify();
            
            console.log('✅ Apple Music Service initialized');
            return true;
        } catch (error) {
            console.warn('⚠️ Apple Music Service initialization failed:', error.message);
            // Continue with fallback method
            return false;
        }
    }

    /**
     * Load private key from Netlify function
     */
    async loadPrivateKeyFromNetlify() {
        try {
            let headers = {};
            if (window.authService?.isAuthenticated) {
                const token = await window.authService.client.getTokenSilently();
                headers['Authorization'] = `Bearer ${token}`;
            }

            const response = await fetch('/.netlify/functions/get-apple-music-key', { headers });
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }

            const data = await response.json();
            this.credentials.privateKey = data.privateKey;
            
            console.log('🔑 Apple Music private key loaded from Netlify');
        } catch (error) {
            throw new Error(`Failed to load Apple Music private key: ${error.message}`);
        }
    }

    /**
     * Get or generate a valid Apple Music developer token
     */
    async getDeveloperToken() {
        // Check if current token is still valid
        if (this.token && this.tokenExpiry && Date.now() < this.tokenExpiry - 300000) { // 5 min buffer
            return this.token;
        }

        // Generate new token
        if (!this.credentials.privateKey) {
            const initialized = await this.initialize();
            if (!initialized) {
                throw new Error('Apple Music Service not properly initialized');
            }
        }

        try {
            const jwt = new AppleMusicJWT(
                this.credentials.teamId,
                this.credentials.keyId,
                this.credentials.privateKey
            );

            this.token = await jwt.generateToken(24); // 24 hour expiration
            this.tokenExpiry = Date.now() + (23 * 60 * 60 * 1000); // 23 hours

            console.log('🎵 New Apple Music developer token generated');
            return this.token;
        } catch (error) {
            throw new Error(`Failed to generate Apple Music token: ${error.message}`);
        }
    }

    /**
     * Test Apple Music API connection
     */
    async testConnection() {
        try {
            const token = await this.getDeveloperToken();
            return await AppleMusicAuth.testConnection(token);
        } catch (error) {
            throw new Error(`Apple Music connection test failed: ${error.message}`);
        }
    }

    /**
     * Search Apple Music catalog
     */
    async search(query, types = ['songs'], limit = 10) {
        try {
            const token = await this.getDeveloperToken();
            const typesParam = types.join(',');
            
            const response = await fetch(
                `https://api.music.apple.com/v1/catalog/us/search?term=${encodeURIComponent(query)}&types=${typesParam}&limit=${limit}`,
                {
                    headers: {
                        'Authorization': `Bearer ${token}`,
                        'Content-Type': 'application/json'
                    }
                }
            );

            if (!response.ok) {
                throw new Error(`Search failed: ${response.status} - ${response.statusText}`);
            }

            return await response.json();
        } catch (error) {
            throw new Error(`Apple Music search failed: ${error.message}`);
        }
    }

    /**
     * Get album details
     */
    async getAlbum(albumId) {
        try {
            const token = await this.getDeveloperToken();
            
            const response = await fetch(
                `https://api.music.apple.com/v1/catalog/us/albums/${albumId}`,
                {
                    headers: {
                        'Authorization': `Bearer ${token}`,
                        'Content-Type': 'application/json'
                    }
                }
            );

            if (!response.ok) {
                throw new Error(`Get album failed: ${response.status} - ${response.statusText}`);
            }

            return await response.json();
        } catch (error) {
            throw new Error(`Failed to get album: ${error.message}`);
        }
    }
}

// Create global Apple Music service instance
window.appleMusicService = new AppleMusicService();

// Auto-initialize when config is ready
window.addEventListener('configReady', async () => {
    try {
        await window.appleMusicService.initialize();
    } catch (error) {
        console.warn('Failed to auto-initialize Apple Music service:', error.message);
    }
});

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { AppleMusicJWT, AppleMusicAuth, AppleMusicService };
}